autode/ext/src/idpp.cpp

```diff
@@ -0,0 +1,912 @@
+        arrx::array1d final_ds = arrx::zeros(n_bonds);
+
+        size_t counter = 0;
+        for (int atom_i = 0; atom_i < n_atoms; atom_i++) {
+            for (int atom_j = 0; atom_j < n_atoms; atom_j++) {
+                if (atom_i >= atom_j) continue;
+
+                auto coord_i = arrx::slice(
+                    init_coords, atom_i * 3, atom_i * 3 + 3
+                );
+                auto coord_j = arrx::slice(
+                    init_coords, atom_j * 3, atom_j * 3 + 3
+                );
+                double dist = arrx::norm_l2(coord_i - coord_j);
+                init_ds[counter] = dist;
+
+                auto coord_i_2 = arrx::slice(
+                    final_coords, atom_i * 3, atom_i * 3 + 3
+                );
+                auto coord_j_2 = arrx::slice(
+                    final_coords, atom_j * 3, atom_j * 3 + 3
+                );
+                dist = arrx::norm_l2(coord_i_2 - coord_j_2);
+                final_ds[counter] = dist;
+                counter++;
+            }
+        }
+
+        // interpolate internal coordinates (distances)
+        this->all_target_ds.clear();
+        for (int k = 0; k < num_images; k++) {
+            // S_k = S_init + (S_fin - S_init) * k / (n-1)
+            double factor = static_cast<double>(k)
+                            / static_cast<double>(num_images - 1);
+            arrx::array1d target_ds = init_ds + (final_ds - init_ds) * factor;
+            this->all_target_ds.push_back(target_ds);
+        }
+    }
+
+    void IDPPPotential::calc_idpp_engrad(const int idx, Image& img) const {
+        /* Calculate the IDPP energy/gradient for the
+         * supplied image
+         *
+         * Arguments:
+         *   idx: The index of the image
+         *
+         *   img: The image for which to calculate the energy/grad.
+         *        The image object is modified in-place
+         */
+        ensure(idx >= 0 && idx < n_images, "Index out of bounds");
+        ensure(img.coords.size() == n_atoms * 3
+               && img.grad.size() == img.coords.size(),
+            "Provided image does not have correct number of atoms");
+
+        if (idx == 0 || idx == n_images - 1) return;
+
+        img.en = 0.0;
+        img.grad.fill(0.0);
+
+        // pointer to items of target_ds[idx]
+        auto target_d_ptr = all_target_ds[idx].begin();
+
+        arrx::array1d dist_vec;
+        for (int atom_i = 0; atom_i < n_atoms; atom_i++) {
+            for (int atom_j = 0; atom_j < n_atoms; atom_j++) {
+                if (atom_i >= atom_j) continue;
+                auto coord_i = arrx::slice(
+                    img.coords, atom_i * 3, atom_i * 3 + 3
+                );
+                auto coord_j = arrx::slice(
+                    img.coords, atom_j * 3, atom_j * 3 + 3
+                );
+                arrx::noalias(dist_vec) = coord_i - coord_j;
+                double dist = arrx::norm_l2(dist_vec);
+                img.en += 1.0 / std::pow(dist, 4)
```

Comment:
```suggestion
                auto dist_pow2 = dist * dist
                auto dist_pow4 = dist_pow2 * dist_pow2
                auto dist_pow5 = dist_pow4 * dist
                auto dist_pow6 = dist_pow5 * dist
                img.en += 1.0 / dist_pow4
```
is probably faster (then use it below too)